src/base/static/libs/maps/mapboxgl-provider.js

```diff
@@ -22,17 +137,75 @@ export default (container, options) => {
     new mapboxgl.NavigationControl(options.control),
     options.control.position,
   );
+  const addLayers = layers => {
+    layers = Array.isArray(layers) ? layers : [layers];
+    layers.forEach(layer => map.addLayer(layer));
+  };
 
   return AbstractMapFactory({
     on: (event, callback, context) => {
       callback = callback.bind(context);
       map.on(event, callback);
     },
 
+    off: (event, callback) => {
+      map.off(event, callback);
+    },
+
+    onLayerEvent: (event, layerId, callback, context) => {
+      callback = callback.bind(context);
+      map.on(event, layerId, callback);
+    },
+
+    bindPlaceLayerEvent: (eventName, layerId, callback) => {
+      layersCache[layerId].forEach(mapProviderLayer => {
```

Comment:
Are we assuming that `layersCache[layerId]` will always be an array? For `createRasterTileLayer` and `createWMTSLayer`, it adds an object to `layersCache[layerId]`. Are we assuming those layer types will never be bound to an event? If so, then why are we storing those layers in `layersCache`?

I also wonder whether we even need `layersCache` - are we not able to fetch the Map Provider Layers from the map using their id's? If so, we should be able to pass in a Layer to `bindPlaceLayerEvent`, regenerate the Map Provider Id(s) from the Layer, and have the map fetch the Map Provider Layers from those id's.

Also, I think it would be  better practice to pass in the Layer to `bindPlaceLayerEvent` anyway, at least to get in the habit. Layer is first class, not `layer.id`.

Reply:
It's true that for Mapbox we don't need to cache the map provider layer representations. I think saving just the generated ids would be sufficient. 

I suppose we could do away with the cache entirely. My concern. though, is that we do a lot of work to generate the internal map provider ids from the first-class representation. We could probably abstract the id generation from the other work done in the `configRuleToXLayer` routines to save some work, but why not just perform that work once and cache the result? Otherwise we'll need to recreate all internal ids on every layer event-- a click, a mouseenter, a mouseout, a visibility trigger, etc. I see the argument for keeping our generated layers "dumb" and making the first-class representation the source of truth, but I also think there could be a place for caching intermediate work.

Reply:
Ah ok - caching the id's, or the Map Provider Layers, sounds reasonable for now.